### PR TITLE
Hide notifications

### LINKED ISSUE #368: Automatically hide informative notifications after some time
for informative ones, check the interface in general. @OlliKiljunen has a good vision on this

### PR DESCRIPTION
**Pull request by jaakkonarhi**
# Description of the PR

Automatic hiding for notifications. Closes #368

Maybe I'm hiding too many notifications. Error notifications get hidden by default so they don't need to be changed

# Testing
<table>
  <tr>
    <th>unit</th>
    <th>integration</th>
    <th>manual</th>
  </tr>
  <tr>
    <td>
      <ul>
        <li>- [ ] new <b>unit</b> tests created</li>
        <li>- [ ] all <b>unit</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [ ] new <b>integration</b> tests created</li>
        <li>- [x] all <b>integration</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [x] <b>manual</b> testing went well</li>
      </ul>
    </td>
  </tr>
</table>  
  
# Have you updated the [TESTING.md](/TESTING.md) or other relevant documentation?

- [ ] Yes
- [ ] Not yet. I will do it next.
- [x] Not relevant

### CODE REVIEW

**Comment by OlliKiljunen** [comment]
>Error notifications get hidden by default so they don't need to be changed
AFAIK, they still stay in the notification list. Not all of them should IMO, for example, "Module is missing" notification that occurs when trying to submit an assignment whose module is not yet loaded.

File: src/main/java/fi/aalto/cs/apluscourses/intellij/notifications/NotificationUtil.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  }
+
+  /**
+   * Shows a notification and hides it after 6 seconds.
+   *
+   * @param notification The Notification to be shown.
+   * @param project      The Project where the Notification is shown.
+   */
+  public static void notifyAndHide(@NotNull Notification notification, @Nullable Project project) {
+    // IntelliJ implementation of notifyAndHide is missing the project parameter from notify
```
Comment (OlliKiljunen) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/notifications/NotificationUtil.java:25]: For reference: https://youtrack.jetbrains.com/issue/IDEA-259792

**Review by superseacat — COMMENTED** [review]
the implementation is beautiful 🤩

**Comment by OlliKiljunen** [comment]
To make the build pass, please change the unit tests accordingly so that they verify the correct method calls.

From build log:
```
fi.aalto.cs.apluscourses.intellij.actions.SubmitExerciseActionTest > testSubmitExerciseAction FAILED
    org.mockito.exceptions.verification.WantedButNotInvoked at SubmitExerciseActionTest.java:240

fi.aalto.cs.apluscourses.intellij.actions.SubmitExerciseActionTest > testNotifiesNoExerciseSelected FAILED
    org.mockito.exceptions.verification.WantedButNotInvoked at SubmitExerciseActionTest.java:254

fi.aalto.cs.apluscourses.model.SubmissionStatusUpdaterTest > testSubmissionStatusUpdater FAILED
    org.mockito.exceptions.verification.WantedButNotInvoked at SubmissionStatusUpdaterTest.java:77
```

**Comment by OlliKiljunen** [comment]
Discussion on topic in Slack: https://apluslms.slack.com/archives/CSDV9CA31/p1611057690016500

**Review by OlliKiljunen — APPROVED** [review]
Very good! Nice and clean code. :+1:

File: src/main/java/fi/aalto/cs/apluscourses/intellij/notifications/NotificationUtil.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    Notifications.Bus.notify(notification, project);
+    Executors.newSingleThreadExecutor().submit(() -> {
+      try {
+        Thread.sleep(6000);
+        notification.expire();
```
[the remark this answers is not in the document]
Comment (OlliKiljunen) [reply src/main/java/fi/aalto/cs/apluscourses/intellij/notifications/NotificationUtil.java:30]: It is, as noted by @nikke234. See line 342 in platform's [Notification.java](https://upsource.jetbrains.com/idea-ce/file/idea-ce-4b94ba01122752d7576eb9d69638b6e89d1671b7/platform/platform-api/src/com/intellij/notification/Notification.java).

**Comment by stellatsv** [comment]
The application and the test suite work as expected. The implementation is very comprehensive and clean!

**Review by stellatsv — APPROVED** [review]

**Pull request merged**